eclipse-theia/theia-blueprint
# Enable Portable Mode in Blueprint

## Portable Mode in Theia Blueprint

Creating a `data` folder at the root of a Theia Blueprint application install (next to the executable) will make it portable. So all of the app data and user data will get written to the `data` folder instead of their default locations.

The `.theia-blueprint` folder is moved from the user’s home directory (`%userprofile%` or `~/`) to a folder called `user-data` inside the `data` folder. `user-data` contains all of the settings, preferences and extensions.

In addition, all of the application data which is stored by default inside a `Theia Blueprint` folder in (`%appdata%` or `~/.config` ) is moved to a folder called `app-data` inside the `data` folder. `app-data` contains all of the application’s cache and layout information.

## How to Start (User Guide)

### Windows, Linux, MacOS

Download and run the correct Theia Blueprint for your platform.

During the installation wizard, select the download location for the Blueprint application. After the install is finished, navigate to the download location and create a `data` folder. The file folder should look similar to the following:

```tsx
- TheiaBlueprint (or application name)
	- TheiaBlueprint.exe
	- data
	- ...
```

As such, all of the data created by Theia Blueprint will be stored inside the `data` folder, more specifically inside `user-data` and `app-data` . 

### Updating Theia Blueprint Portable

Download the latest Theia Blueprint version and run the installer. Copy and paste your current `data` folder to the root of the directory where you installed the latest version of Blueprint. This should update your application.

### Migrate to Portable Mode

If you want to make an existing installation of Blueprint portable, you can do so with the following steps:

1. Create a `data` folder at the root of your Blueprint install directory
2. Copy the following and paste it into a subfolder inside `data` called `user-data`:
    - Windows: `%userprofile%/.theia-blueprint`
    - Linux: `~/.theia-blueprint`
    - macOS: `~/.theia`
3. Copy the following and paste it into a subfolder inside `data` called `app-data`:
    - Windows: `%appdata%/Theia Blueprint`
    - Linux: `$HOME/.config/Theia Blueprint`
    - macOS: `$HOME/Library/Application Support/Theia Blueprint`

This should make your existing installation portable.

## How to test

The following changes only makes the `user-data` portable (settings, preferences, recent workspace ...)
For the `app-data` to be made portable, the following PR should be merged, published to npm and the dependencies updated:

You can test that the portable-mode is functional by packaging a new Theia Blueprint application using `yarn` and then `yarn electron package`. Then, you can run the executable and create a `data` folder at the download location. This should write all the user data from `~/.theia-blueprint` to `data/user-data`.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
